File: csharp/1012.numbers-with-repeated-digits.cs
/*
 * @lc app=leetcode id=1012 lang=csharp
 *
 * [1012] Numbers With Repeated Digits
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static int NumPermutation(int len, int free, IDictionary<(int Len, int Free), int> memo) {
        if (len == 0) {
            return 1;
        }
        var key = (len: len, Free: free);
        if (memo.TryGetValue(key, out int ret)) {
            return ret;
        }
        return memo[key] = NumPermutation(len - 1, free, memo) * (free - len + 1);
    }

    public int NumDupDigitsAtMostN(int N) {
        var list = new List<int>();
        for (int i = N + 1; i > 0; i /= 10) {
            list.Add(i % 10);
        }
        int len = list.Count, distinct = 0;
        var memo = new Dictionary<(int Len, int Free), int>();
        var used = new HashSet<int>();
        for (int i = 1; i < len; ++i) {
            distinct += 9 * NumPermutation(i - 1, 9, memo);
        }
        for (int i = 0; i < len; ++i) {
            int curr = list[len - i - 1];
            for (int j = i == 0 ? 1 : 0; j < curr; ++j) {
                if (!used.Contains(j)) {
                    distinct += NumPermutation(len - i - 1, 9 - i, memo);
                }
            }
            if (!used.Add(curr)) {
                break;
            }
        }
        return N - distinct;
    }
}
// @lc code=end

File: java/1012.numbers-with-repeated-digits.java
/*
 * @lc app=leetcode id=1012 lang=java
 *
 * [1012] Numbers With Repeated Digits
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start

class Solution {
    public int numDupDigitsAtMostN(int n) {
        String s = Integer.toString(n + 1);
        int len = s.length(), distinct = 0;
        Map<Key, Integer> memo = new HashMap<>();
        for (int l = 1; l < len; ++l) {
            distinct += countPermutation(l - 1, 9, memo) * 9;
        }
        int used = 0;
        for (int i = 0; i < len; ++i) {
            int c = s.charAt(i) - '0';
            for (int start = i == 0 ? 1 : 0; start < c; ++start) {
                if (((1 << start) & used) == 0) {
                    distinct += countPermutation(len - i - 1, 9 - i, memo);
                }
            }
            if (((1 << c) & used) != 0) {
                break;
            }
            used |= 1 << c;
        }
        return n - distinct;
    }

    private static int countPermutation(int len, int free, Map<Key, Integer> memo) {
        if (len == 0) {
            return 1;
        }
        Key key = new Key(len, free);
        Integer cached = memo.get(key);
        if (cached != null) {
            return cached;
        }
        int v = countPermutation(len - 1, free, memo) * (free - (len - 1));
        memo.put(key, v);
        return v;
    }

    private record Key(int len, int free) {}
}
// @lc code=end

